File: packages/fundamentals/charts/native/kotlin/com/pyreon/runtime/PyreonChartWebView.kt
package com.pyreon.runtime

import org.json.JSONArray
import org.json.JSONObject
import org.json.JSONTokener

data class PyreonChartWebViewSelection(
    val seriesName: String? = null,
    val seriesIndex: Int? = null,
    val name: String? = null,
    val dataIndex: Int? = null,
    val value: Any? = null,
    val componentType: String? = null,
)

data class PyreonChartWebViewEvent(val name: String, val payload: Map<String, Any?>)
data class PyreonChartWebViewError(val message: String)

fun pyreonChartWebViewData(
    option: String,
    commands: String,
    loading: Boolean,
    loadingOptions: String,
    group: String? = null,
): String = try {
    JSONObject()
        .put("__pyreonChartHost", 1)
        .put("option", JSONTokener(option).nextValue())
        .put("commands", JSONArray(commands))
        .put("loading", JSONObject().put("visible", loading).put("options", JSONObject(loadingOptions)))
        .apply { if (!group.isNullOrEmpty()) put("group", group) }
        .toString()
} catch (_: Exception) {
    option
}

fun pyreonDispatchChartWebViewMessage(
    message: String,
    onSelect: ((PyreonChartWebViewSelection) -> Unit)? = null,
    onEvent: ((PyreonChartWebViewEvent) -> Unit)? = null,
    onError: ((PyreonChartWebViewError) -> Unit)? = null,
) {
    val raw = try { JSONTokener(message).nextValue() } catch (_: Exception) {
        onSelect?.invoke(PyreonChartWebViewSelection(name = message))
        return
    }
    if (raw !is JSONObject) return
    if (raw.has("error")) {
        onError?.invoke(PyreonChartWebViewError(raw.opt("error")?.toString() ?: "Hosted chart failed"))
        return
    }
    if (raw.optInt("__pyreonChartEvent") == 1) {
        onEvent?.invoke(PyreonChartWebViewEvent(
            name = raw.optString("name"),
            payload = raw.optJSONObject("payload")?.toMap() ?: emptyMap(),
        ))
        return
    }
    onSelect?.invoke(PyreonChartWebViewSelection(
        seriesName = raw.optString("seriesName").ifEmpty { null },
        seriesIndex = raw.optInt("seriesIndex").takeIf { raw.has("seriesIndex") },
        name = raw.optString("name").ifEmpty { null },
        dataIndex = raw.optInt("dataIndex").takeIf { raw.has("dataIndex") },
        value = raw.opt("value").takeUnless { it == JSONObject.NULL },
        componentType = raw.optString("componentType").ifEmpty { null },
    ))
}

private fun JSONObject.toMap(): Map<String, Any?> = keys().asSequence().associateWith { key ->
    when (val value = opt(key)) {
        JSONObject.NULL -> null
        is JSONObject -> value.toMap()
        is JSONArray -> (0 until value.length()).map { index -> value.opt(index) }
        else -> value
    }
}

File: packages/fundamentals/charts/native/swift/PyreonChartWebView.swift
import Foundation

public struct PyreonChartWebViewSelection {
    public let seriesName: String?
    public let seriesIndex: Int?
    public let name: String?
    public let dataIndex: Int?
    public let value: Any?
    public let componentType: String?
}

public struct PyreonChartWebViewEvent {
    public let name: String
    public let payload: [String: Any]
}

public struct PyreonChartWebViewError: Error {
    public let message: String
}

public func pyreonChartWebViewData(
    option: String,
    commands: String,
    loading: Bool,
    loadingOptions: String,
    group: String? = nil
) -> String {
    let decodedOption = pyreonChartDecode(option) ?? [:]
    let decodedCommands = pyreonChartDecode(commands) as? [Any] ?? []
    let decodedLoadingOptions = pyreonChartDecode(loadingOptions) as? [String: Any] ?? [:]
    var envelope: [String: Any] = [
        "__pyreonChartHost": 1,
        "option": decodedOption,
        "commands": decodedCommands,
        "loading": ["visible": loading, "options": decodedLoadingOptions],
    ]
    if let group, !group.isEmpty { envelope["group"] = group }
    guard let data = try? JSONSerialization.data(withJSONObject: envelope),
          let result = String(data: data, encoding: .utf8)
    else { return option }
    return result
}

public func pyreonDispatchChartWebViewMessage(
    _ message: String,
    onSelect: ((PyreonChartWebViewSelection) -> Void)? = nil,
    onEvent: ((PyreonChartWebViewEvent) -> Void)? = nil,
    onError: ((PyreonChartWebViewError) -> Void)? = nil
) {
    guard let raw = pyreonChartDecode(message) else {
        onSelect?(PyreonChartWebViewSelection(
            seriesName: nil, seriesIndex: nil, name: message, dataIndex: nil,
            value: nil, componentType: nil))
        return
    }
    guard let object = raw as? [String: Any] else { return }
    if let error = object["error"] {
        onError?(PyreonChartWebViewError(message: error as? String ?? String(describing: error)))
        return
    }
    if (object["__pyreonChartEvent"] as? NSNumber)?.intValue == 1 {
        onEvent?(PyreonChartWebViewEvent(
            name: object["name"] as? String ?? "",
            payload: object["payload"] as? [String: Any] ?? [:]))
        return
    }
    onSelect?(PyreonChartWebViewSelection(
        seriesName: object["seriesName"] as? String,
        seriesIndex: (object["seriesIndex"] as? NSNumber)?.intValue,
        name: object["name"] as? String,
        dataIndex: (object["dataIndex"] as? NSNumber)?.intValue,
        value: object["value"],
        componentType: object["componentType"] as? String))
}

private func pyreonChartDecode(_ json: String) -> Any? {
    guard let data = json.data(using: .utf8) else { return nil }
    return try? JSONSerialization.jsonObject(with: data, options: [.fragmentsAllowed])
}
